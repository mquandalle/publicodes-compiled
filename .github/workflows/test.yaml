name: test
on: push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: oven-sh/setup-bun@v1
      - run: bun install
      - run: bun test
