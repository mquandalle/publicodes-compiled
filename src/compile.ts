import { walk } from "zimmerframe";
import { builders as b } from "estree-toolkit";
import { generate } from "astring";
import { type ASTPublicodesNode } from "./parse";

export function compile(parsedPublicodes: ASTPublicodesNode) {
  const state = { injectedRuntime: new Set() };
  const esTree = walk(parsedPublicodes, state, {
    publicodes(node, { visit }) {
      return b.objectExpression(
        node.rules.map((ruleNode) =>
          b.property("init", b.literal(ruleNode.name), visit(ruleNode))
        )
      );
    },

    rule(node, { visit }) {
      return b.arrowFunctionExpression([], visit(node.value));
    },

    constant(node) {
      return b.literal(node.value);
    },

    reference(node) {
      return b.callExpression(b.identifier("r"), [b.literal(node.name)]);
    },

    operation(node, { visit }) {
      return b.binaryExpression(
        node.operator,
        visit(node.left),
        visit(node.right)
      );
    },

    produit(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.assiette,
        right: node.taux,
      });
    },

    unitConversion(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.factor,
        right: node.value,
      });
    },

    barÃ¨me(node, { state, visit }) {
      state.injectedRuntime.add(
        "const evalBareme = (a, t) => t.map((u, i) => ((Math.max(Math.min(a, u.plafond ?? Infinity) - (t[i - 1]?.plafond ?? 0), 0)) * u.taux)).reduce((a, b) => a + b, 0);"
      );
      return b.callExpression(b.identifier("evalBareme"), [
        visit(node.assiette),
        b.arrayExpression(
          node.tranches.map((t) =>
            b.objectExpression(
              ["taux", t.plafond && "plafond"]
                .filter(Boolean)
                .map((attr) =>
                  b.property("init", b.identifier(attr), visit(t[attr]))
                )
            )
          )
        ),
      ]);
    },
  });

  const jsCodeRules = generate(esTree);
  const injectedRuntime = Array.from(state.injectedRuntime).join(";\n");
  return `((r) => {${injectedRuntime}\nreturn ${jsCodeRules}})`;
}
