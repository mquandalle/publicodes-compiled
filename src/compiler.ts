import { walk } from "zimmerframe";
import { builders as b } from "estree-toolkit";
import { generate } from "astring";
import { type ASTPublicodesNode } from "./parser";

export function compile(parsedPublicodes: ASTPublicodesNode) {
  const state = { injectedRuntime: new Set() };
  const esTree = walk(parsedPublicodes, state, {
    publicodes(node, { visit }) {
      return b.objectExpression(
        node.rules.map((ruleNode) =>
          b.property("init", b.literal(ruleNode.name), visit(ruleNode))
        )
      );
    },

    rule(node, { visit }) {
      return b.arrowFunctionExpression([b.identifier("r")], visit(node.value));
    },

    constant(node) {
      return b.literal(node.value);
    },

    reference(node) {
      return b.callExpression(b.identifier("r"), [b.literal(node.name)]);
    },

    operation(node, { visit }) {
      return b.binaryExpression(
        node.operator,
        visit(node.left),
        visit(node.right)
      );
    },

    undefined() {
      return b.identifier("undefined");
    },

    produit(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.assiette,
        right: node.taux,
      });
    },

    unitConversion(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.factor,
        right: node.value,
      });
    },

    barÃ¨me(node, { state, visit }) {
      state.injectedRuntime.add(
        "const evalBareme = (a, t) => t.map((u, i) => ((Math.max(Math.min(a, u.plafond ?? Infinity) - (t[i - 1]?.plafond ?? 0), 0)) * u.taux)).reduce((a, b) => a + b, 0);"
      );
      return b.callExpression(b.identifier("evalBareme"), [
        visit(node.assiette),
        b.arrayExpression(
          node.tranches.map((t) =>
            b.objectExpression(
              ["taux", t.plafond && "plafond"]
                .filter(Boolean)
                .map((attr) =>
                  b.property("init", b.identifier(attr), visit(t[attr]))
                )
            )
          )
        ),
      ]);
    },

    ["une de ces conditions"](node, { visit }) {
      let res = visit(node.value[0]);
      for (let i = 1; i < node.value.length; i++) {
        res = b.logicalExpression("||", res, visit(node.value[i]));
      }
      return res;
    },

    ["toutes ces conditions"](node, { visit }) {
      let res = visit(node.value[0]);
      for (let i = 1; i < node.value.length; i++) {
        res = b.logicalExpression("&&", res, visit(node.value[i]));
      }
      return res;
    },

    variations(node, { visit }) {
      return b.conditionalExpression(
        visit(node.value[0].si),
        visit(node.value[0].alors),
        visit(node.value[1].sinon)
      );
    },

    ["applicable si"](node, { visit }) {
      return b.conditionalExpression(
        visit(node["applicable si"]),
        visit(node["value"]),
        b.identifier("null")
      );
    },
  });

  const jsCodeRules = generate(esTree);
  const injectedRuntime = Array.from(state.injectedRuntime).join(";\n");
  return `(() => {${injectedRuntime}\nreturn ${jsCodeRules}})`;
}
