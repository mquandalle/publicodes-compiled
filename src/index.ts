import { parse } from "./parser";
import { compile } from "./compiler";

export class CompiledPublicodes {
  compiledRules: any;
  cache: any;
  constructor(source: string) {
    const jsCode = compile(parse(source));
    const r = (x: string) => this.evaluate(x);
    this.compiledRules = eval(jsCode)(r);
    this.cache = {};
  }

  evaluate(ruleName: string) {
    this.cache[ruleName] ??= this.compiledRules[ruleName]();
    return this.cache[ruleName];
  }

  resetCache() {
    this.cache = {};
  }
}
