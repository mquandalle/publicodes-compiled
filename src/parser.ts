import { tokenize, type Token } from "./tokenizer";
import { link } from "./linker";

const mechanismSignatures = {
  "une de ces conditions": [],
  "toutes ces conditions": [],
  variations: [],
  plafond: [],
  "applicable si": [],
  somme: [],
  unité: [],
  "par défaut": [],
  valeur: [],
  produit: [],
  barème: ["assiette", "tranches"],
};

const mechanismNames = Object.keys(mechanismSignatures);

export const keysWithExpression = new Set([
  ...mechanismNames,
  ...Object.values(mechanismSignatures).flat(),
  ..."abattement, si, alors, sinon, applicable si, assiette, avec, commune, grille, intercommunalité, nom, non applicable si, par défaut, par, plafond, plancher, possibilités, produit, règle, remplace, rend non applicable, sauf dans, somme, taux, toutes ces conditions, tranches, une de ces conditions, unité, valeur, variations".split(
    ", "
  ),
]);

const chainableMecanisms = [
  "par défaut",
  "unité",
  "applicable si",
  "plancher",
  "plafond",
];

const textFields = [
  "description",
  "titre",
  "remplace",
  "lien",
  "type",
  "question",
  "nom",
];

export type ASTNode =
  | { type: "constant"; value: boolean | number | string; unit?: string }
  | { type: "undefined" }
  | { type: "operation"; operator: string; left: ASTNode; right: ASTNode }
  | { type: "reference"; name: string }
  | { type: "produit"; assiette: ASTNode; taux: ASTNode }
  | {
      type: "barème";
      assiette: ASTNode;
      tranches: Array<{ taux: ASTNode; plafond?: ASTNode }>;
    };

export type ASTRuleNode = { type: "rule"; name: string; value: ASTNode };

export type ASTPublicodesNode = {
  type: "publicodes";
  rules: Array<ASTRuleNode>;
};

export function parse(
  source: string,
  { withLoc = false, availableRules = [] } = {}
): ASTPublicodesNode {
  const tokens: Token[] = tokenize(source, { withLoc });
  const parsedRules: ASTRuleNode[] = [];
  let currentRuleNode;
  let index = 0;

  while (index < tokens.length) {
    const currentToken = tokens[index];
    if (currentToken.type === "outdent") {
      index++;
      continue;
    } else if (currentToken.type === "key") {
      parseRule();
    } else {
      throw new Error(`Unexpected token ${currentToken.type}`);
    }
  }

  function parseRule() {
    const keyToken = tokens[index++];
    const ruleName = keyToken.value;
    // a new rule, this one is empty
    if (index >= tokens.length || tokens[index]?.type === "key") {
      parsedRules.push({
        type: "rule",
        name: ruleName,
        value: { type: "undefined" },
      });
    } else {
      currentRuleNode = { type: "rule", name: ruleName };
      currentRuleNode.value = parseExpression();
      if (withLoc) {
        currentRuleNode.start = keyToken.start;
        currentRuleNode.end = currentRuleNode.value.end;
      }
      parsedRules.push(currentRuleNode);
    }
  }

  function parseExpression() {
    if (tokens[index]?.type === "indent") {
      index++; // skip the 'indent' token
      const node =
        tokens[index]?.type === "list-item" ? parseList() : parseRecord();
      if (tokens[index]?.type === "outdent") {
        index++;
      }
      return node;
    }
    return parseInlineExpression();
  }

  function parseList() {
    const list = [];
    while (index < tokens.length && tokens[index].type === "list-item") {
      index++;

      if (tokens[index].type === "key") {
        list.push(parseRecord());
      } else {
        list.push(parseExpression());
      }
    }
    return list;
  }

  function parseRecord() {
    const entries = [];
    let start = tokens[index].start,
      end = tokens[index].end;
    while (index < tokens.length && tokens[index].type === "key") {
      end = tokens[index].end;
      const key = tokens[index++].value;
      if (textFields.includes(key)) {
        const nextToken = tokens[index];
        if (
          nextToken.type === "text" ||
          nextToken.type === "reference" ||
          nextToken.type === "number"
        ) {
          currentRuleNode[key] = nextToken.value;
          index++;
        } else if (nextToken.type === "key") {
          currentRuleNode[key] = "";
        } else if (nextToken.type === "indent") {
          currentRuleNode[key] = parseExpression();
        } else {
          throw new Error(`Unexpected token ${JSON.stringify(tokens[index])}`);
        }
      } else {
        if (key === "unité") {
          entries.push([key, tokens[index++]]);
        } else {
          entries.push([key, parseExpression()]);
        }
      }
    }
    if (entries.length === 0) {
      throw new Error(`Unexpected token ${JSON.stringify(tokens[index])}`);
    }
    if (
      index < tokens.length &&
      tokens[index].type !== "outdent" &&
      tokens[index].type !== "list-item"
    ) {
      throw new Error(`Expected token at ${index}`);
    }

    let node;

    const mechanismsEntries = entries.filter(([key]) =>
      mechanismNames.includes(key)
    );
    if (mechanismsEntries.length !== entries.length) {
      node = { ...Object.fromEntries(entries), type: "record" };
      if (withLoc) {
        node.start = start;
        node.end = end;
      }
      return node;
    }
    const nonChainableMechanism = mechanismsEntries.filter(
      ([key]) => !chainableMecanisms.includes(key)
    );
    const chainableMechanisms = mechanismsEntries.filter(([key]) =>
      chainableMecanisms.includes(key)
    );

    if (nonChainableMechanism.length > 1) {
      throw new Error(`Unexpected token ${JSON.stringify(tokens[index])}`);
    }

    const mechanismName = nonChainableMechanism[0]?.[0];
    const value = nonChainableMechanism[0]?.[1];
    const mechanismKeys = mechanismSignatures[mechanismName];
    if (!mechanismName) {
      node = { type: "constant", value: undefined };
    } else if (mechanismName === "valeur") {
      node = value;
    } else if (mechanismKeys.length === 0) {
      node = { type: mechanismName, value };
    } else {
      node = { ...value, type: mechanismName };
    }
    for (let chainableMechanism of chainableMechanisms) {
      const mechanismName = chainableMechanism[0];
      node = {
        type: mechanismName,
        [mechanismName]: chainableMechanism[1],
        value: node,
      };
    }

    if (withLoc) {
      return { ...node, start, end };
    }
    return node;
  }

  function parseInlineExpression(): ASTNode {
    return parseComparison();
  }

  function parseComparison(): ASTNode {
    let node = parseAdditionSubstration();
    let start = node.start;
    let end = node.end;

    let token = tokens[index];
    if (
      token &&
      token.type === "operator" &&
      (token.value === "=" ||
        token.value === ">=" ||
        token.value === ">" ||
        token.value === "<" ||
        token.value === "<=")
    ) {
      index++;
      const right = parseAdditionSubstration();
      end = right.end;
      node = {
        type: "operation",
        operator: token.value === "=" ? "==" : token.value,
        left: node,
        right,
      };
    }
    if (withLoc) {
      return { ...node, start, end };
    }
    return node;
  }

  function parseAdditionSubstration(): ASTNode {
    let node = parseMultiplicationDivision();
    let start = node.start;
    let end = node.end;

    while (index < tokens.length) {
      const token = tokens[index];
      if (
        token.type === "operator" &&
        (token.value === "+" || token.value === "-")
      ) {
        index++;
        const right = parseMultiplicationDivision();
        end = right.end;
        node = {
          type: "operation",
          operator: token.value,
          left: node,
          right,
        };
      } else {
        break;
      }
    }
    if (withLoc) {
      return { ...node, start, end };
    }
    return node;
  }

  function parseMultiplicationDivision(): ASTNode {
    let node = parseParenthesizedExpression();
    let start = node.start;
    let end = node.end;

    while (index < tokens.length) {
      const token = tokens[index];
      if (
        token.type === "operator" &&
        (token.value === "*" || token.value === "/")
      ) {
        index++;
        const right = parseParenthesizedExpression();
        end = right.end;
        node = {
          type: "operation",
          operator: token.value,
          left: node,
          right,
        };
      } else {
        break;
      }
    }
    if (withLoc) {
      return { ...node, start, end };
    }
    return node;
  }

  function parseParenthesizedExpression(): ASTNode {
    if (tokens[index].type === "paren-open") {
      let start = tokens[index].start;
      index++;
      const node = parseInlineExpression();
      let end = tokens[index].end;
      if (tokens[index++].type !== "paren-close") {
        throw new Error(`Unexpected token ${JSON.stringify(tokens[index])}`);
      }
      if (withLoc) {
        return { ...node, start, end };
      }
      return node;
    }
    return parseTerminalNode();
  }

  function parseTerminalNode(): ASTNode {
    const token = tokens[index++];
    let node;
    let start = token.start;
    let end = token.end;
    if (token.type === "number") {
      if (tokens[index]?.type === "unit") {
        end = tokens[index].end;
        node = {
          type: "constant",
          value: token.value,
          unit: tokens[index++].value,
        };
      } else {
        node = { type: "constant", value: token.value, unit: "" };
      }
    } else if (token.type === "reference") {
      node = { type: "reference", name: token.value };
    } else if (token.type === "text") {
      node = { type: "constant", value: token.value };
    } else if (token.type === "boolean") {
      node = { type: "constant", value: token.value };
    } else {
      throw new Error(
        `Unexpected token ${JSON.stringify(token)} at ${index - 1}`
      );
    }

    if (withLoc) {
      return { ...node, start, end };
    } else {
      return node;
    }
  }

  const parsedProgram = { type: "publicodes", rules: parsedRules } as const;

  if (withLoc) {
    parsedProgram.start = parsedRules[0].start;
    parsedProgram.end = parsedRules[parsedRules.length - 1].end;
  }

  // return parsedProgram;
  return link(parsedProgram, availableRules);
}

export function parseJsObject(jsObj, options) {
  // HACK
  return parse(
    Object.entries(jsObj)
      .map(([k, v]) => `${k}: ${v}`)
      .join("\n"),
    options
  );
}
