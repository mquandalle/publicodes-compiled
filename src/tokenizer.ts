import { keysWithExpression } from "./parser";

export type Token =
  | { type: "number"; value: number }
  | { type: "boolean"; value: boolean }
  | { type: "text"; value: string }
  | { type: "operator"; value: string }
  | { type: "reference"; value: string }
  | { type: "unit"; value: string }
  | { type: "key"; value: string }
  | { type: "paren-open" }
  | { type: "paren-close" }
  | { type: "indent" }
  | { type: "outdent" }
  | { type: "list-item" };

const ruleNameRegex = /[\p{L}$](([$\p{L}\x20\.'0-9]|\S-\S)*[\p{L}$0-9])?/uy;
const booleanRegex = /(oui|non)/y;
const numberRegex = /[0-9]+(\.[0-9]+)?/y;
const unitRegex = /[a-z€]+(\/[a-z€]+)?/y;
const infixOperatorRegex = /(<=|>=|<|>|=|\+|\-|\*|\/)/y;
const endOfLine = /.*/y;
const spaceRegex = /\x20+/y;
const startMultiLineStringRegex = /\x20*\|\x20*\n/y;

type TokenWithLoc = Token & { start?: number; end?: number };

export function tokenize(
  source: string,
  { withLoc = false } = {}
): TokenWithLoc[] {
  let tokens: Token[] = [];
  let currentIndent = 0;
  let indentStack: number[] = [];
  let cursor = 0;
  let prevCursorPos = 0;

  function matchRegex(regex: RegExp): string | undefined {
    regex.lastIndex = cursor;
    const matched = regex.exec(source);
    if (matched) {
      if (withLoc) {
        prevCursorPos = cursor;
      }
      cursor += matched[0].length;
      return matched[0];
    }
  }
  function matchSingleChar<const C extends string>(char: C): C | undefined {
    const matched = source[cursor] === char;
    if (matched) {
      if (withLoc) {
        prevCursorPos = cursor;
      }
      cursor++;
      return char;
    }
  }
  function push(token: Token) {
    if (withLoc) {
      tokens.push({ ...token, start: prevCursorPos, end: cursor });
    } else {
      tokens.push(token);
    }
  }

  while (cursor < source.length) {
    if (cursor === 0 || matchRegex(/\n+/y)) {
      let indent = matchRegex(spaceRegex)?.length || 0;
      let isListItem = false;
      if (matchSingleChar("-")) {
        indent++;
        indent += matchRegex(spaceRegex)?.length || 0;
        isListItem = true;
      } else if (matchSingleChar("#")) {
        matchRegex(endOfLine);
        continue;
      }

      if (indent > currentIndent) {
        push({ type: "indent" });
        indentStack.push(currentIndent);
        currentIndent = indent;
      }

      while (indent < currentIndent) {
        push({ type: "outdent" });
        currentIndent = indentStack.pop() || 0;
      }
      if (isListItem) {
        push({ type: "list-item" });
      }

      if (cursor > 0) {
        continue;
      }
    }

    if (matchSingleChar("#")) {
      matchRegex(endOfLine);
      continue;
    }

    if (matchRegex(spaceRegex)) {
      continue;
    }

    const operatorToken = matchRegex(infixOperatorRegex);
    if (operatorToken) {
      push({ type: "operator", value: operatorToken });
      continue;
    }

    if (matchSingleChar("(")) {
      push({ type: "paren-open" });
      continue;
    }

    if (matchSingleChar(")")) {
      push({ type: "paren-close" });
      continue;
    }

    let quoteType = matchSingleChar('"') || matchSingleChar("'");
    if (quoteType) {
      let value = "";
      let legacyDuplicatedQuotes = false;

      const invertQuote = (q: '"' | "'") => (q === '"' ? "'" : '"');
      if (matchSingleChar(invertQuote(quoteType))) {
        legacyDuplicatedQuotes = true;
        quoteType = invertQuote(quoteType);
      }

      for (; cursor < source.length; cursor++) {
        if (source[cursor] === "\\") {
          value += source[++cursor];
          continue;
        } else if (source[cursor] === quoteType) {
          push({ type: "text", value });
          if (withLoc) {
            tokens[tokens.length - 1].end += 1;
          }
          cursor++;

          if (legacyDuplicatedQuotes) {
            if (!matchSingleChar(invertQuote(quoteType))) {
              throw new Error("Unterminated legacy string");
            }
          }
          break;
        }

        value += source[cursor];
      }

      if (cursor > source.length) {
        throw new Error("Unterminated string");
      }

      continue;
    }
    const refToken = matchRegex(ruleNameRegex);
    if (refToken) {
      matchRegex(spaceRegex);
      if (matchSingleChar(":")) {
        push({ type: "key", value: refToken });
        if (matchRegex(startMultiLineStringRegex)) {
          multiLineString();
        } else if (refToken === "unité") {
          matchRegex(spaceRegex);
          const unit = matchRegex(unitRegex);
          if (unit) {
            push({ type: "unit", value: unit });
          } else {
            throw new Error("Missing unit");
          }
        } else if (
          !keysWithExpression.has(refToken) &&
          indentStack.length > 0
        ) {
          matchRegex(spaceRegex);
          const text = matchRegex(endOfLine);
          if (text) {
            push({ type: "text", value: text });
          }
        }
      } else {
        if (refToken.match(/^(oui|non)$/)) {
          push({ type: "boolean", value: refToken === "oui" });
        } else {
          push({ type: "reference", value: refToken });
        }
      }
      continue;
    }

    const booleanToken = matchRegex(booleanRegex);
    if (booleanToken) {
      push({ type: "boolean", value: booleanToken === "oui" });
      continue;
    }

    const numberToken = matchRegex(numberRegex);
    if (numberToken) {
      matchRegex(spaceRegex);
      const percentage = matchSingleChar("%");
      const value = parseFloat(numberToken);
      if (percentage) {
        push({ type: "number", value: value / 100 });
      } else {
        push({ type: "number", value });
      }
      const unit = matchRegex(unitRegex);
      if (unit) {
        push({ type: "unit", value: unit });
      }
      continue;
    }
    throw prettyPrintSyntaxError(`Cannot tokenize`);
  }

  function multiLineString() {
    const endMultiLineString = RegExp(
      "(.+?)\\n\\x20{0," + currentIndent + "}([^\\x20\\n]|$)",
      "ys"
    );
    endMultiLineString.lastIndex = cursor;
    const multilineString = endMultiLineString.exec(source);

    if (!multilineString) {
      throw new Error("Unterminated multi-line string");
    }
    const length = multilineString[1].length;
    const value = source
      .slice(cursor, cursor + length)
      .replace(/(^|\n)\x20*/g, "\n")
      .trim();
    cursor += multilineString[1].length;
    push({ type: "text", value });
  }

  function prettyPrintSyntaxError(message = "Syntax error") {
    const error = new Error(message);
    const lines = source.split("\n");
    let prevCount = 0;
    let count = 0;
    const lineWithCursorIndex = lines.findIndex((line, i) => {
      count += line.length + 1;
      if (cursor < count && cursor > prevCount) {
        return true;
      }
      prevCount = count;
    });

    const printLine = (start: number) => (line: string, i: number) =>
      `${start + i + 1} | ${line}`;

    const startLine = Math.max(0, lineWithCursorIndex - 3);
    const endLine = Math.min(lines.length, lineWithCursorIndex + 3);
    error.message += `\n${lines
      .slice(startLine, lineWithCursorIndex + 1)
      .map(printLine(startLine))
      .join("\n")}`;
    error.message +=
      "\n" +
      " ".repeat(lineWithCursorIndex.toString().length + 1) +
      "|" +
      " ".repeat(cursor - prevCount - 1) +
      "⬆️";
    error.message += `\n${lines
      .slice(lineWithCursorIndex + 1, endLine)
      .map(printLine(lineWithCursorIndex + 1))
      .join("\n")}`;
    return error;
  }

  return tokens;
}

export function printTokens(tokens: Token[], { lineNumbers = false } = {}) {
  return tokens
    .map(
      (t, i) =>
        `${
          lineNumbers
            ? i.toString().padEnd(tokens.length.toString().length + 1, " ")
            : ""
        }${t.type}${"value" in t ? `(${t.value})` : ""}`
    )
    .join("\n");
}
