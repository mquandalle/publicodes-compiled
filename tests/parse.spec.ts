import { expect, test } from "bun:test";
import { parse } from "../src/parser";
import { printTokens, tokenize } from "../src/tokenizer";

test("parse list and records", () => {
  const listAndRecords = `
simple record:
    a: 1
    b: 2

simple lists:
    a:
        - "ok"
        - "okok"
    b:
        - "ok"
        - "okok"

imbricated records:
    a:
        b:
            c: 1

mixed list and records:
    a:
        a1:
            - "ok"
            - "okok"
        a2:
            - "ok"
            - "okok"
    b:
        b1:
            - b11:
                - "ok"
                - b111:
                    - "ok"
                    - "okok"
        b2: okok

recursive lists:
    a:
      - toutes ces conditions:
        - "ok"
        - "okok"
      - une de ces conditions:
        - "ok"
        - "okok"

deep tree:
    a:
        aL:
            - a:
                aL:
                    - a:
                        aOk: ok
    b: ok
`.trim();
  const parsedPublicodes = parse(listAndRecords);

  expect(
    printTokens(tokenize(listAndRecords), { lineNumbers: true })
  ).toMatchSnapshot();

  const rules = Object.fromEntries(
    parsedPublicodes.rules.map((r) => [r.name, r.value])
  );

  expect(rules["simple record"]).toEqual({
    type: "record",
    a: { type: "constant", value: "1" },
    b: { type: "constant", value: "2" },
  });

  expect(rules["simple lists"]).toEqual({
    type: "record",
    a: [
      { type: "constant", value: "ok" },
      { type: "constant", value: "okok" },
    ],
    b: [
      { type: "constant", value: "ok" },
      { type: "constant", value: "okok" },
    ],
  });

  expect(rules["imbricated records"]).toEqual({
    type: "record",
    a: {
      type: "record",
      b: {
        type: "record",
        c: { type: "constant", value: "1" },
      },
    },
  });

  expect(rules["recursive lists"]).toEqual({
    type: "record",
    a: [
      {
        type: "toutes ces conditions",
        value: [
          { type: "constant", value: "ok" },
          { type: "constant", value: "okok" },
        ],
      },
      {
        type: "une de ces conditions",
        value: [
          { type: "constant", value: "ok" },
          { type: "constant", value: "okok" },
        ],
      },
    ],
  });

  expect(rules["mixed list and records"]).toEqual({
    type: "record",
    a: {
      type: "record",
      a1: [
        { type: "constant", value: "ok" },
        { type: "constant", value: "okok" },
      ],
      a2: [
        { type: "constant", value: "ok" },
        { type: "constant", value: "okok" },
      ],
    },
    b: {
      type: "record",
      b1: [
        {
          type: "record",
          b11: [
            { type: "constant", value: "ok" },
            {
              type: "record",
              b111: [
                { type: "constant", value: "ok" },
                { type: "constant", value: "okok" },
              ],
            },
          ],
        },
      ],
      b2: { type: "constant", value: "okok" },
    },
  });

  expect(rules["deep tree"]).toEqual({
    type: "record",
    a: {
      type: "record",
      aL: [
        {
          type: "record",
          a: {
            type: "record",
            aL: [
              {
                type: "record",
                a: {
                  type: "record",
                  aOk: { type: "constant", value: "ok" },
                },
              },
            ],
          },
        },
      ],
    },
    b: { type: "constant", value: "ok" },
  });
});

test("valeur", () => {
  const parsedPublicodes = parse(
    `
aide:
    titre: Aide vélo
    valeur: 50 €
    `.trim()
  );
  expect(parsedPublicodes.rules[0]).toEqual({
    type: "rule",
    name: "aide",
    titre: "Aide vélo",
    unit: {
      type: "number",
      unit: "€",
    },
    value: {
      type: "constant",
      value: 50,
      unit: "€",
    },
  });
});

test("chained mechanisms", () => {
  const parsedPublicodes = parse(
    `
aide:
    applicable si: oui
    valeur: 50 €
`.trim()
  );
  expect(parsedPublicodes.rules[0]).toEqual({
    type: "rule",
    name: "aide",
    unit: {
      type: "number",
      unit: "€",
    },
    value: {
      type: "applicable si",
      "applicable si": {
        type: "constant",
        value: true,
      },
      value: {
        type: "constant",
        value: 50,
        unit: "€",
      },
    },
  });
});
