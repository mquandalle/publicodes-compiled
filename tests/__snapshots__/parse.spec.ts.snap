// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse list and records 1`] = `
"0   key(simple record)
1   indent
2   key(a)
3   text(1)
4   key(b)
5   text(2)
6   outdent
7   key(simple lists)
8   indent
9   key(a)
10  indent
11  list-item
12  text(ok)
13  list-item
14  text(okok)
15  outdent
16  key(b)
17  indent
18  list-item
19  text(ok)
20  list-item
21  text(okok)
22  outdent
23  outdent
24  key(imbricated records)
25  indent
26  key(a)
27  indent
28  key(b)
29  indent
30  key(c)
31  text(1)
32  outdent
33  outdent
34  outdent
35  key(mixed list and records)
36  indent
37  key(a)
38  indent
39  key(a1)
40  indent
41  list-item
42  text(ok)
43  list-item
44  text(okok)
45  outdent
46  key(a2)
47  indent
48  list-item
49  text(ok)
50  list-item
51  text(okok)
52  outdent
53  outdent
54  key(b)
55  indent
56  key(b1)
57  indent
58  list-item
59  key(b11)
60  indent
61  list-item
62  text(ok)
63  list-item
64  key(b111)
65  indent
66  list-item
67  text(ok)
68  list-item
69  text(okok)
70  outdent
71  outdent
72  outdent
73  key(b2)
74  text(okok)
75  outdent
76  outdent
77  key(recursive lists)
78  indent
79  key(a)
80  indent
81  list-item
82  key(toutes ces conditions)
83  indent
84  list-item
85  text(ok)
86  list-item
87  text(okok)
88  outdent
89  list-item
90  key(une de ces conditions)
91  indent
92  list-item
93  text(ok)
94  list-item
95  text(okok)
96  outdent
97  outdent
98  outdent
99  key(deep tree)
100 indent
101 key(a)
102 indent
103 key(aL)
104 indent
105 list-item
106 key(a)
107 indent
108 key(aL)
109 indent
110 list-item
111 key(a)
112 indent
113 key(aOk)
114 text(ok)
115 outdent
116 outdent
117 outdent
118 outdent
119 outdent
120 key(b)
121 text(ok)"
`;
